// zmp-ui/cjs/components/header/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _clsx = _interopRequireDefault(require("clsx"));

var _reactRouterDom = require("react-router-dom");

var _class = require("../../utils/class");

var _icon = _interopRequireDefault(require("../icon"));

var _button = _interopRequireDefault(require("../button"));

/**
 * Header component: Component Header cho Mini App, có thể sử dụng thay thế cho Header mặc định
 *
 * @category general
 * @subCategory Header
 * @component
 *
 * @example
import React from 'react';
import { Header, Page } from 'zmp-ui'

export default () => {
    return (
        <Page>
            <Header title="Zalo Mini App" />
        </Page>
    );
}
 */
var Header = /*#__PURE__*/_react["default"].forwardRef(function (props, ref) {
  var navigate = (0, _reactRouterDom.useNavigate)();
  var _props$showBackIcon = props.showBackIcon,
      showBackIcon = _props$showBackIcon === void 0 ? true : _props$showBackIcon,
      backIcon = props.backIcon,
      onBackClick = props.onBackClick,
      title = props.title,
      className = props.className,
      style = props.style,
      id = props.id;
  var prefixCls = (0, _class.getPrefixCls)("header");
  var cls = (0, _clsx["default"])(prefixCls, {}, className);

  var handleBackClick = function handleBackClick(event) {
    if (onBackClick) {
      onBackClick(event);
      return;
    }

    event.preventDefault();
    event.stopPropagation();
    navigate(-1);
  };

  var backIconEl = backIcon || /*#__PURE__*/_react["default"].createElement(_icon["default"], {
    icon: "zi-chevron-left"
  });

  return /*#__PURE__*/_react["default"].createElement("div", {
    ref: ref,
    className: cls,
    style: style,
    id: id
  }, showBackIcon && /*#__PURE__*/_react["default"].createElement("span", {
    className: prefixCls + "-back"
  }, /*#__PURE__*/_react["default"].createElement(_button["default"], {
    className: prefixCls + "-back-btn",
    type: "neutral",
    variant: "tertiary",
    onClick: handleBackClick,
    icon: backIconEl
  })), /*#__PURE__*/_react["default"].createElement("span", {
    className: prefixCls + "-title"
  }, title));
});

Header.displayName = "zaui-header";
var _default = Header;
exports["default"] = _default;
