// zmp-ui/esm/components/header/index.js
import React from "react";
import classNames from "clsx";
import { useNavigate } from "react-router-dom";
import { getPrefixCls } from "../../utils/class";
import Icon from "../icon";
import Button from "../button";
/**
 * Header component: Component Header cho Mini App, có thể sử dụng thay thế cho Header mặc định
 *
 * @category general
 * @subCategory Header
 * @component
 *
 * @example
import React from 'react';
import { Header, Page } from 'zmp-ui'

export default () => {
    return (
        <Page>
            <Header title="Zalo Mini App" />
        </Page>
    );
}
 */

var Header = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var navigate = useNavigate();
  var _props$showBackIcon = props.showBackIcon,
      showBackIcon = _props$showBackIcon === void 0 ? true : _props$showBackIcon,
      backIcon = props.backIcon,
      onBackClick = props.onBackClick,
      title = props.title,
      className = props.className,
      style = props.style,
      id = props.id;
  var prefixCls = getPrefixCls("header");
  var cls = classNames(prefixCls, {}, className);

  var handleBackClick = function handleBackClick(event) {
    if (onBackClick) {
      onBackClick(event);
      return;
    }

    event.preventDefault();
    event.stopPropagation();
    navigate(-1);
  };

  var backIconEl = backIcon || /*#__PURE__*/React.createElement(Icon, {
    icon: "zi-chevron-left"
  });
  return /*#__PURE__*/React.createElement("div", {
    ref: ref,
    className: cls,
    style: style,
    id: id
  }, showBackIcon && /*#__PURE__*/React.createElement("span", {
    className: prefixCls + "-back"
  }, /*#__PURE__*/React.createElement(Button, {
    className: prefixCls + "-back-btn",
    type: "neutral",
    variant: "tertiary",
    onClick: handleBackClick,
    icon: backIconEl
  })), /*#__PURE__*/React.createElement("span", {
    className: prefixCls + "-title"
  }, title));
});
Header.displayName = "zaui-header";
export default Header;
