// zmp-ui/esm/components/router/AnimationRoutes.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
var _excluded = ["element"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

import React, { useState, useMemo, useRef, useEffect, cloneElement, Children, createRef } from "react";
import classNames from "clsx";
import { NavigationType, useNavigationType, useLocation, Routes } from "react-router-dom";
import { CSSTransition, TransitionGroup } from "react-transition-group";
import { getPrefixCls } from "../../utils/class";
import { iOS } from "../../utils/device";
import Context from "./context";
/**
 * AnimationRoutes component: Wrapper thay thế cho Routes của react-router, component này thêm hiệu ứng khi chuyển trang
 *
 * @category Router
 * @catPosition 2
 * @subcategory AnimationRoutes
 * @component
 * @typedef {object} AnimationRoutes
 * @prop {ReactNode} children Các Route page
 *
 *
 * @example <data>{"live": false}</data>
 function MyApp(props){
    return (
        <App>
            <ZMProuter>
                 <AnimationRoutes>
                     <Route path="/" element={<HomePage />} />
                     <Route path="/page1" element={<Page1 />} />
                </AnimationRoutes>
            </ZMPRouter>
        </App>
    );
}
 */

var AnimationRoutes = function AnimationRoutes(props) {
  var _classNames;

  var _useState = useState(true),
      animate = _useState[0],
      setAnimate = _useState[1];

  var navigationType = useNavigationType();
  var children = props.children;
  var location = useLocation();
  var pathname = location.pathname;
  var initialX = useRef(null);
  var initialY = useRef(null);
  var nodeRefs = useRef({});
  var direction = useRef("");
  var prefixCls = getPrefixCls("routes");

  var getKey = function getKey(path) {
    if (!path) return "";
    var pathToCompare = path.charAt(0) !== "/" ? "/" + path : path;

    if (pathToCompare.lastIndexOf("/") > 0) {
      pathToCompare = pathToCompare.substring(0, pathToCompare.length - 1);
    }

    return pathToCompare;
  };

  var routList = useMemo(function () {
    return Children.map(children, function (child) {
      if ( /*#__PURE__*/React.isValidElement(child)) {
        var _child$props = child.props,
            element = _child$props.element,
            restProps = _objectWithoutPropertiesLoose(_child$props, _excluded);

        nodeRefs.current[getKey(child.props.path)] = /*#__PURE__*/createRef();
        if (!element || element.props.replace === true) return child;
        return _objectSpread(_objectSpread({}, child), {}, {
          props: _objectSpread(_objectSpread({}, restProps), {}, {
            element: element
          })
        });
      }

      return child;
    });
  }, [children]);

  var startTouch = function startTouch(e) {
    initialX.current = e.touches[0].clientX;
    initialY.current = e.touches[0].clientY;
  };

  var moveTouch = function moveTouch(e) {
    if (initialX.current === null) {
      return;
    }

    if (initialY.current === null) {
      return;
    }

    var currentX = e.touches[0].clientX;
    var currentY = e.touches[0].clientY;
    var diffX = initialX.current - currentX;
    var diffY = initialY.current - currentY;

    if (initialX.current <= 24 && Math.abs(diffX) > Math.abs(diffY) && diffX <= 0 && animate) {
      e.preventDefault();
      setAnimate(false);
      e.preventDefault();
    }

    initialX.current = null;
    initialY.current = null;
  };

  useEffect(function () {
    var isIOS = iOS();

    if (isIOS) {
      document.addEventListener("touchstart", startTouch, {
        passive: false
      });
      document.addEventListener("touchmove", moveTouch, {
        passive: false
      });
    }

    return function () {
      if (isIOS) {
        document.removeEventListener("touchstart", startTouch);
        document.removeEventListener("touchmove", moveTouch);
      }
    };
  }, []);
  var contextValue = useMemo(function () {
    return {
      animate: animate
    };
  }, [animate]);

  if (animate) {
    direction.current = navigationType === NavigationType.Pop ? "backward" : "forward";
  }

  var animationCls = classNames((_classNames = {}, _classNames[prefixCls + "-" + direction.current] = animate, _classNames[prefixCls + "-no-animation"] = !animate, _classNames));
  var wrapperCls = classNames(prefixCls, {});
  var nodeRef = nodeRefs.current[getKey(pathname)];
  return /*#__PURE__*/React.createElement(Context.Provider, {
    value: contextValue
  }, /*#__PURE__*/React.createElement(TransitionGroup, {
    className: wrapperCls,
    childFactory: function childFactory(child) {
      return /*#__PURE__*/cloneElement(child, {
        classNames: animationCls,
        timeout: animate ? 400 : 200
      });
    }
  }, /*#__PURE__*/React.createElement(CSSTransition, {
    key: pathname,
    timeout: animate ? 400 : 200,
    onExited: function onExited() {
      setAnimate(true);
    },
    nodeRef: nodeRef,
    unmountOnExit: true
  }, /*#__PURE__*/React.createElement("div", {
    ref: nodeRef,
    className: prefixCls + "-item"
  }, /*#__PURE__*/React.createElement(Routes, {
    location: location
  }, routList)))));
};

AnimationRoutes.displayName = "zaui-animation-router";
export default AnimationRoutes;
